File: Myrtille.Common/Network/Buffer.cs
/*
    Myrtille: A native HTML4/5 Remote Desktop Protocol client.

    Copyright(c) 2014-2021 Cedric Coste

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Myrtille.Network
{
    public class DataBuffer<T>
    {
        public int Size { get; set; }
        public int Delay { get; set; }

        private List<T> _data;
        private object _dataLock;

        public delegate void FlushDelegate(List<T> data);
        public FlushDelegate SendBufferData { get; set; }

        private bool _enabled;
        private CancellationTokenSource _timeout;

        public DataBuffer(int size, int delay)
        {
            Size = size;
            Delay = delay;

            _data = new List<T>();
           _dataLock = new object();
        }

        public void Start()
        {
            if (_enabled)
                return;

            _enabled = true;

            _timeout = new CancellationTokenSource();
            Task.Delay(Delay > 0 ? Delay : 1, _timeout.Token).ContinueWith(task =>
            {
                Flush();
            }, TaskContinuationOptions.NotOnCanceled);
        }

        public void Stop()
        {
            _enabled = false;
        }

        public void AddItem(T item)
        {
            if (!_enabled)
                return;

            lock (_dataLock)
            {
                _data.Add(item);

                if (_data.Count == 1)
                {
                    if (_timeout != null)
                    {
                        _timeout.Cancel();
                        _timeout = null;
                    }

                    _timeout = new CancellationTokenSource();
                    Task.Delay(Delay > 0 ? Delay : 1, _timeout.Token).ContinueWith(task =>
                    {
                        Flush();
                    }, TaskContinuationOptions.NotOnCanceled);
                }
                else if (_data.Count >= Size)
                {
                    Flush();
                }
            }
        }

        public void Flush()
        {
            if (!_enabled)
                return;

            if (_timeout != null)
            {
                _timeout.Cancel();
                _timeout = null;
            }

            lock (_dataLock)
            {
                if (_data.Count > 0)
                {
                    SendBufferData?.Invoke(_data);
                    _data.Clear();
                }
            }

            _timeout = new CancellationTokenSource();
            Task.Delay(Delay > 0 ? Delay : 1, _timeout.Token).ContinueWith(task =>
            {
                Flush();
            }, TaskContinuationOptions.NotOnCanceled);
        }
    }
}

File: Myrtille.Web.Blazor/wwwroot/js/network/buffer.js
/*
    Myrtille: A native HTML4/5 Remote Desktop Protocol client.

    Copyright(c) 2014-2021 Cedric Coste

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*****************************************************************************************************************************************************************************************************/
/*** Buffer                                                                                                                                                                                        ***/
/*****************************************************************************************************************************************************************************************************/

function Buffer(base, config, dialog, network)
{
    var bufferSize = config.getBufferSize();

    var bufferDelay = config.getBufferDelayBase();
    this.getBufferDelay = function() { return bufferDelay; };
    this.setBufferDelay = function(delay) { bufferDelay = delay; };

    var bufferData = new Array();

    var sendEmptyBuffer = false;
    this.getSendEmptyBuffer = function() { return sendEmptyBuffer; };

    var clearBuffer = true;
    this.setClearBuffer = function(enabled) { clearBuffer = enabled; };

    var bufferTimeout = null;
    
    this.init = function()
    {
        try
        {
            // if using xhr only, updates are returned within the user inputs xhr responses; thus, the display is not updated while the user does nothing (no keyboard or mouse event)
            // to work that out, the user inputs buffer should be sent *even if empty* in order to allow retrieving the latest updates
            // problem is, on a low latency network, this will generate alot of xhr trafic which may interfere with others xhr's such as periodical and adaptative fullscreen updates!
            // to avoid that, and because there is no need for a frantic polling, a small timing is added
            if (config.getNetworkMode() == config.getNetworkModeEnum().XHR)
            {
                bufferDelay = config.getBufferDelayEmpty();
                sendEmptyBuffer = true;
            }

            bufferTimeout = window.setTimeout(function() { doFlush(); }, bufferDelay);
        }
        catch (exc)
        {
            dialog.showDebug('buffer init error: ' + exc.message);
            config.setBufferEnabled(false);
        }
    };

    this.addItem = function(data)
    {
        bufferData.push(data);

        if (bufferData.length == 1)
        {
            if (bufferTimeout != null)
            {
                window.clearTimeout(bufferTimeout);
                bufferTimeout = null;
            }

            //dialog.showDebug('buffer delay: ' + bufferDelay);
            bufferTimeout = window.setTimeout(function() { doFlush(); }, bufferDelay);
        }
        else if (bufferData.length >= bufferSize)
        {
            //dialog.showDebug('buffer is full, flushing');
            doFlush();
        }
    };

    this.flush = function()
    {
        doFlush();
    };

    function doFlush()
    {
        try
        {
            if (!config.getBufferEnabled())
                return;

            if (bufferTimeout != null)
            {
                window.clearTimeout(bufferTimeout);
                bufferTimeout = null;
            }

            if (bufferData.length > 0 || sendEmptyBuffer)
            {
                //dialog.showDebug('flushing buffer: ' + bufferData.toString());

                // send the buffered data over the network
                network.send(bufferData.toString());

                // if the data was sent successfully, clear the buffer
                if (clearBuffer)
                {
                    //dialog.showDebug('clearing buffer');
                    bufferData = [];
                }
            }

            // set the next buffer flush tick
            //dialog.showDebug('buffer delay: ' + bufferDelay);
            bufferTimeout = window.setTimeout(function() { doFlush(); }, bufferDelay);
        }
        catch (exc)
        {
            dialog.showDebug('buffer flush error: ' + exc.message);
        }
    }
}